spec.html

```diff
@@ -30493,24 +30492,20 @@ <h2>Syntax</h2>
           &lt;ZWJ&gt;
 
         RegExpUnicodeEscapeSequence[U] ::
-          [+U] `u` LeadSurrogate `\u` TrailSurrogate
-          [+U] `u` LeadSurrogate
-          [+U] `u` TrailSurrogate
-          [+U] `u` NonSurrogate
+          [+U] RegExpUnicodeSurrogatePair
+          [+U] [lookahead &notin; RegExpUnicodeSurrogatePair] `u` Hex4Digits
```

Comment:
`lookahead &notin; RegExpUnicodeSurrogatePair` seems a little suspicious to me, since `RegExpUnicodeSurrogatePair` describes a set of sequences of terminals rather than just a set of terminals. As far as I know the only other place where there are sequences of length greater than 1 in a lookahead-restriction-set is the `async function` restriction in [ExpressionStatement](https://tc39.es/ecma262/#sec-expression-statement), and there the sequence is of length precisely two (and also it's written out more explicitly).

Could this instead be written as

```
[+U] `u` LeadSurrogate [lookahead ≠ `\u` TrailSurrogate]
[+U] `u` LeadSurrogate `\u` TrailSurrogate
[+U] `u` TrailSurrogate
[+U] `u` NonSurrogate
```
? That feels clearer to me, if it's equivalent. (And if it's not, I'm confused.)

Reply:
That is equivalent, though still a lookahead of six code points. And not more clear to me, but I would be willing to switch to it if there's consensus.

Reply:
Currently, in the ES grammars, a lookahead-constraint either:
(a) occurs at the end of a right-hand-side, or
(b) occurs before a nonterminal, where that nonterminal derives phrases that begin with the disallowed sequences (i.e. the constraint is 'scoped' to that nonterminal).

So the right-hand-side:
```
[lookahead &notin; RegExpUnicodeSurrogatePair] `u` Hex4Digits
```
would be quite unusual, in that the lookahead-constraint has to "look through" the terminal `u` and nonterminal _HexDigits_, and then look past them to a potential `u` _Hex4Digits_ following. For this reason, I prefer @bakkot's suggestion:
```
LeadSurrogate [lookahead != `\u` TrailSurrogate]
```
as it eliminates the "look through" and winds up with a fairly standard end-of-RHS constraint. (But yes, the nature of its lookahead-sequence would require tweaking 5.1.5 Grammar Notation.)

Also, I think I'd prefer it to come *after* the Lead+Trail right-hand-side:
```
[+U] `u` LeadSurrogate `\u` TrailSurrogate
[+U] `u` LeadSurrogate [lookahead != `\u` TrailSurrogate]
```
(The other thing I like about this solution is that just those two lines make it really obvious why the lookahead-constraint is needed.)